src/Database/Esqueleto/PostgreSQL/JSON.hs

```diff
@@ -0,0 +1,618 @@
+  --     |         | Key/value pairs are matched based on their key value.                  |
+  --  -  | integer | Delete the array element with specified index (Negative integers count | '["a", "b"]'::jsonb - 1
+  --     |         | from the end). Throws an error if top level container is not an array. |
+  --  #- | text[]  | Delete the field or element with specified path                        | '["a", {"b":1}]'::jsonb #- '{1,b}'
+  --     |         | (for JSON arrays, negative integers count from the end)                |
+  --
+  -- /Works with any PostgreSQL of version >= 10/
+  --
+  --     | Type    | Description                                                            |  Example
+  -- ----+---------+------------------------------------------------------------------------+-------------------------------------------------
+  --  -  | text[]  | Delete multiple key/value pairs or string elements from left operand.  | '{"a": "b", "c": "d"}'::jsonb - '{a,c}'::text[]
+  --     |         | Key/value pairs are matched based on their key value.                  |
+  -- @
+  , (||.)
+  , (-.)
+  , (--.)
+  , (#-.)
+  ) where
+
+#if __GLASGOW_HASKELL__ < 804
+import Data.Semigroup
+#endif
+import Data.Aeson (encode, eitherDecodeStrict)
+import qualified Data.Aeson as Aeson (Value)
+import qualified Data.ByteString.Lazy as BSL (toStrict)
+import Data.Text (Text)
+import qualified Data.Text as T (concat, pack)
+import qualified Data.Text.Encoding as TE (decodeUtf8, encodeUtf8)
+import Database.Esqueleto.Internal.Language hiding ((?.), (-.), (||.))
+import Database.Esqueleto.Internal.PersistentImport
+import Database.Esqueleto.Internal.Sql
+
+
+infixl 6 ->., ->>., #>., #>>.
+infixl 6 @>., <@., ?., ?|., ?&.
+infixl 6 ||., -., --., #-.
+
+
+-- | This function extracts the jsonb value from a JSON array or object,
+-- depending on whether you use an @int@ (@Left i@ in Haskell) or a
+-- @text@ (@Right t@ in Haskell).
+--
+-- As long as the left operand is @jsonb@, this function will not
+-- throw an exception, but will return @NULL@ when an @int@ is used on
+-- anything other than a JSON array, or a @text@ is used on anything
+-- other than a JSON object.
+--
+-- === __PostgreSQL Documentation__
+--
+-- /Works with any PostgreSQL of version >= 9.3/
+--
+-- @
+--     | Type | Description                                |  Example                                         | Example Result
+-- ----+------+--------------------------------------------+--------------------------------------------------+----------------
+--  -> | int  | Get JSON array element (indexed from zero) | '[{"a":"foo"},{"b":"bar"},{"c":"baz"}]'::json->2 | {"c":"baz"}
+--  -> | text | Get JSON object field by key               | '{"a": {"b":"foo"}}'::json->'a'                  | {"b":"foo"}
+-- @
+--
+-- @since 3.1.0
+(->.) :: SqlExpr (Value (Maybe Aeson.Value))
+      -> Either Int Text
```

Comment:
I think the API would be more flexible if we didn't use Haskell values directly, but instead used `SqlExpr` of these values. That's how we're doing things at Lumi right now. It's easy to go from `Text` to `SqlExpr (Value Text)` with `val`, but it's IIRC not possible to go the other way.

If we do that, though, then we need to use something other than `Either`, because (IIRC) there's no easy way to go from an eg `SqlExpr (Value Int)` to a `SqlExpr (Value (Either Int Text))`. The library sometimes defines classes like `SqlString` to allow folks to use all kinds of stuff, and I've used that before to good luck. Something like `class JsonAccessor a` with a few instances for `Int`, `Text`, `String`, `Integer`, etc would be a good way to be more flexible, at the cost of type inference. Eg, `foo ->. val 3` would give an ambiguous type warning, and `foo ->. val "hello"` would give an ambiguous warning with `OverloadedStrings` enabled.

On the other hand, we don't actually, uh, use the `SqlExpr` power *anywhere* in the Lumi codebase that I can tell. All uses of our `|->` operator have a `val "some string literal"` form.

Maybe we should have a `data JsonAccessor = JInt Int | JStr Text` that has `Num` and `IsString` instances :joy:

Reply:
The use case is slim, I think. We also never use a column as the right hand operand. It's almost always taking the JSON value of a column and either filtering on it, or changing it's shape.
Though just to be able to do it, would be nice, in case someone has a specific situation where they, for example, want to filter a row based on the interaction between two columns?

Or maybe go for the data type `data JsonAccessor` and have helper functions that accept any valid value?

`numToJsonAccessor :: Num a => SqlExpr (Value a) -> SqlExpr (Value JsonAccessor)`
`stringToJsonAccessor :: IsString a => SqlExpr (Value a) -> SqlExpr (Value JsonAccessor)`

---

Also:
Any right hand operand (RHO) can be `NULL` in any of the operators. The result will just be `NULL`. If we use `SqlExpr`s, it might be good to just default to `SqlExp (Value (Maybe a))`, where a is the intended type of the RHO.